rfcs/convergence/theme-css-variable-insertion.md

```diff
@@ -0,0 +1,157 @@
+
+## Problem Statement
+
+### Inline styles (Current solution)
+
+Simply rendering all css variables as inline styles onto the DOM element
+
+**_CSS variables can only be used by DOM children - not good in the case of React portals_**
+
+```tsx
+<div style={{...cssVars}}>
+    // happy can inherit CSS variables
+</div>
+
+// React.createPortal()
+// no theme values here ⚠
+<div id="tooltip" />
+```
+
+To get around this we would need to render a `ThemeProvider` for each `Portal`
+
+```tsx
+<Portal>
+    <ThemeProvider> // writes CSS vars inline again
+        <div id="tooltip"> /
+    </ThemeProvider>
+</Portal>
+```
+
+## Detailed Design or Proposal
+
+### Possible solutions
+
+#### Inserting styles to `document.head`
+
+Each `ThemeProvider` will create a style tag and insers all css vars as a single CSS rule for a CSS class unique to that `ThemeProvider`
+
+The classname will be passed down through context to children and can be useful for Portals which can't consume CSS variables from DOM parents if `Inline` insertion is used
+
+```tsx
+const cssVars = themeToCssVars(theme);
+const cssRule = {...};
+// result: .theme-provider { --css-var: '#fff' }
+styleEl.sheet.insertRule(cssRule);
+
+<head>
+  <style id="theme-provider-1" />
+  <style id="theme-provider-2" />
+  <style id="theme-provider-3" />
+</head>
+```
+
+```tsx
+<Portal className="theme-provider-1">
+    <div id="tooltip"> // can use theme \o/
+</Portal>
+```
+
+#### Inheriting parent cssVars
+
+Compares current `Theme` with parent `Theme` and only adds the CSS variables that are different to the parent `Theme`.
+
+Now each subsequent layer of `ThemeProvider` will need to use the CSS class rule of all of its parents.
+
+```tsx
+const cssVars = themeToCssVars(theme);
+const parentCssVars = themeToCssVars(parentTheme);
+const diff = diff(cssVars, parentCssVars)
+const cssRule = {...};
+// result: .theme-provider { --css-var: '#fff' }
+styleEl.sheet.insertRule(cssRule);
+
+<head>
+  <style id="theme-provider-1" />
+  <style id="theme-provider-2" />
+  <style id="theme-provider-3" />
+</head>
+
+
+<div class="theme-provider-1">
+    <div class="theme-provider-1 theme-provider-2">
+        <div class="theme-provider-1 theme-provider-2 theme-provider-3">
+        </div>
+    </div>
+</div>
+```
+
+#### Inheriting parent cssVars and reusing style tag
+
+Same as above except all `ThemeProviders` now share `<style />` tag
+
+```tsx
+const cssVars = themeToCssVars(theme);
+const parentCssVars = themeToCssVars(parentTheme);
+const diff = diff(cssVars, parentCssVars)
+const cssRule = {...};
+// result: .theme-provider { --css-var: '#fff' }
+styleEl.sheet.insertRule(cssRule);
+
+<head>
+  <style id="the one" />
+</head>
+
+
+<div class="theme-provider-1">
+    <div class="theme-provider-1 theme-provider-2">
```

Comment:
I don't understand how this works.
If both theme 1 and theme 2 set the same CSS variable, the one defined later in the DOM wins. How can you control specificity here?

Why are you adding the `theme-provider-1` here? Without the class here, you get the theme 1 CSS variables through cascade and theme 2 vars will override it.

Reply:
> If both theme 1 and theme 2 set the same CSS variable, the one defined later in the DOM wins. How can you control specificity here?

This would be by the order of the `ThemeProvider` components declared. The CSS variables are written during render time. This would result in the variables of theme 2 being written later.

> Why are you adding the theme-provider-1 here? Without the class here, you get the theme 1 CSS variables through cascade and theme 2 vars will override it.

You're right. In the case of a single DOM tree, this is not needed and cascade will definitely do its job. Concatenating classes came from the original motivation to use this for `Portal` which renders out of order, in that case, all classes in the tree would be necessary so that the CSS vars available in the portal is predictable

Reply:
> This would be by the order of the ThemeProvider components declared. The CSS variables are written during render time. This would result in the variables of theme 2 being written later.

I see now - you are creating a unique classname for each `ThemeProvider` instance.

Reply:
Yeah, my bad I'll make it clearer.

But it works for shared classes too, as long as you manage insertion order correctly. Here is an example of that:

https://codesandbox.io/s/wonderful-gould-myrdq?file=/src/App.js